python/lsst/daf/butler/_formatter.py

```diff
@@ -51,12 +63,1034 @@
     from ._dataset_ref import DatasetRef
     from ._dataset_type import DatasetType
     from ._storage_class import StorageClass
-    from .dimensions import DataCoordinate
+    from .datastore.cache_manager import AbstractDatastoreCacheManager
 
     # Define a new special type for functions that take "entity"
     Entity: TypeAlias = DatasetType | DatasetRef | StorageClass | str
 
 
+class FileIntegrityError(RuntimeError):
+    """The file metadata is inconsistent with the metadata supplied by
+    the datastore.
+    """
+
+
+class FormatterNotImplementedError(NotImplementedError):
+    """Formatter does not implement the specific read or write method
+    that is being requested.
+    """
+
+
+class FormatterV2:
+    """Interface for reading and writing datasets using URIs.
+
+    The formatters are associated with a particular `StorageClass`.
+
+    Parameters
+    ----------
+    file_descriptor : `FileDescriptor`, optional
+        Identifies the file to read or write, and the associated storage
+        classes and parameter information.  Its value can be `None` if the
+        caller will never call `Formatter.read` or `Formatter.write`.
+    ref : `DatasetRef`
+        The dataset associated with this formatter. Should not be a component
+        dataset ref.
+    write_parameters : `dict`, optional
+        Any parameters to be hard-coded into this instance to control how
+        the dataset is serialized.
+    write_recipes : `dict`, optional
+        Detailed write Recipes indexed by recipe name.
+    **kwargs
+        Additional arguments that will be ignored but allow for
+        `Formatter` V1 parameters to be given.
```

Comment:
This is because FormatterFactory is creating these instances and it's given `**kwargs` that work for V1 and V2 without it knowing which params are relevant (and that's because the MappingFactory creates the instance and can't know the distinction).